### PR TITLE
Improve resource ref unit tests.

### LINKED ISSUE #5939: Preview fails when resource refs are enabled
From https://github.com/pulumi/pulumi/issues/5933:

Previews fail in very incorrect ways with CLI `2.15.5`.

```ts
import * as aws from "@pulumi/aws";
const bucket = new aws.s3.Bucket("mybucket");
bucket.onObjectCreated("foofunction", async (ev, ctx) => {
    console.log(JSON.stringify(ev));
});
```

```
$ pulumi preview
...
Diagnostics:
  aws:lambda:Permission (foofunction):
    error: aws:lambda/permission:Permission resource 'foofunction' has a problem: "function_name" doesn't comply with restrictions ("^(arn:[\\w-]+:lambda:)?([a-z]{2}-(?:[a-z]+-){1,2}\\d{1}:)?(\\d{12}:)?(function:)?([a-zA-Z0-9-_]+)(:(\\$LATEST|[a-zA-Z0-9-_]+))?$"): "urn:pulumi:dev::promisecapture::aws:s3/bucket:Bucket$aws:s3:BucketEventSubscription$aws:lambda/function:Function::foofunction"
```

This happens when resource refs are enabled.

### PR DESCRIPTION
**Pull request by pgavlin**
- Add component ref coverage to the existing test
- Add coverage for a downlevel SDK communicating with an engine that
  supports resource refs
- Add coverage for a downlevel engine communicating with an SDK that
  supports resource refs

As part of improving coverage, these changes add a knob to explicitly
disable resource refs in the engine without the use of the environment
variable. The environment variable is now only read by the CLI.

I believe that these changes also fix #5939, which I needed to
address in order for the tests to pass.

Contributes to #5943.

### CODE REVIEW

**Comment by pgavlin** [comment]
@justinvp , I think that these changes obviate https://github.com/pulumi/pulumi/pull/5952. I'll add additional SDK-specific tests as a follow-up.

**Comment by justinvp** [comment]
> I think that these changes obviate #5952. I'll add additional SDK-specific tests as a follow-up.

Thanks! Since I'm OOF until Jan, I'm going to close out #5952 and assign #5939 to you.

**Comment by pgavlin** [comment]
Ping. I'd love to get these changes in ASAP to unblock other testing.

File: pkg/cmd/pulumi/util.go
Comment (leezen) [inline pkg/cmd/pulumi/util.go:81]: We can remove the comment, close pulumi-kubernetes#1405, and uncomment the code as well, right?

File: pkg/engine/lifeycletest/resource_reference_test.go
Comment (leezen) [inline pkg/engine/lifeycletest/resource_reference_test.go:174]: I don't think I quite understand this. The resource can be referenced by either the id or the urn? I would have expected this to always be via the urn? Sorry if this is simple, but I'm just not familiar with this part of the code very much.

File: pkg/cmd/pulumi/util.go
Comment (pgavlin) [reply pkg/cmd/pulumi/util.go:81]: > We can remove the comment, close pulumi-kubernetes#1405, and uncomment the code as well, right?

I wasn't certain that was the plan. I don't think that we want to enable resource refs by default until we've landed all of the planned test coverage, do we?

File: pkg/engine/lifeycletest/resource_reference_test.go
Comment (pgavlin) [reply pkg/engine/lifeycletest/resource_reference_test.go:174]: > // TestResourceReferences_DownlevelEngine tests an SDK that supports resource references communicating with an engine
 // that does not.

This test simulates an SDK that supports resource references communicating with an engine that does not support resource references. In that case, the marshaller should transform a resource reference into either a plain URN (for references to component resources) or a plan ID (for references to custom resources). These checks validate that the transformation occurred as expected.

File: pkg/cmd/pulumi/util.go
Comment (leezen) [reply pkg/cmd/pulumi/util.go:81]: Sorry, I didn't mean enable by default, but don't we need some way to do the reverse (enable so we can test)? I assumed that logic would live in this block.

Comment (pgavlin) [reply pkg/cmd/pulumi/util.go:81]: > don't we need some way to do the reverse (enable so we can test)? I assumed that logic would live in this block.

Not for the purposes of these tests, which enable and disable support for resource refs by hitting the engine directly. For E2E/integration tests, though, we do need to restore the `PULUMI_ENABLE_RESOURCE_REFERENCES` envvar. I'll make that change.

**Review by leezen — APPROVED** [review]

**Pull request merged**
